test/extensions/http/cache/file_system_http_cache/file_system_http_cache_test.cc

```diff
@@ -116,12 +121,75 @@ TEST_F(FileSystemHttpCacheTestWithNoDefaultCache, InitialStatsAreSetCorrectly) {
+  env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-b"), file_contents, true);
+  {
+    // TODO(ravenblack): replace this with backdating the files when that's possible.
+    absl::Mutex mu;
+    absl::MutexLock lock(&mu);
+    auto cond = []() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu) { return false; };
+    mu.AwaitWithTimeout(absl::Condition(&cond), absl::Seconds(1));
+  }
+  cache_ = std::dynamic_pointer_cast<FileSystemHttpCache>(
+      http_cache_factory_->getCache(cacheConfig(cfg), context_));
+  waitForEvictionThreadIdle();
+  EXPECT_EQ(cache_->stats().size_bytes_.value(), file_contents.size() * 2);
+  EXPECT_EQ(cache_->stats().size_count_.value(), 2);
+  env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-c"), file_contents, true);
+  env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-d"), file_contents, true);
+  cache_->trackFileAdded(file_contents.size());
+  cache_->trackFileAdded(file_contents.size());
+  waitForEvictionThreadIdle();
+  EXPECT_EQ(cache_->stats().size_bytes_.value(), file_contents.size() * 2);
+  EXPECT_EQ(cache_->stats().size_count_.value(), 2);
+  EXPECT_FALSE(Filesystem::fileSystemForTest().fileExists(absl::StrCat(cache_path_, "cache-a")));
+  EXPECT_FALSE(Filesystem::fileSystemForTest().fileExists(absl::StrCat(cache_path_, "cache-b")));
+  EXPECT_TRUE(Filesystem::fileSystemForTest().fileExists(absl::StrCat(cache_path_, "cache-c")));
+  EXPECT_TRUE(Filesystem::fileSystemForTest().fileExists(absl::StrCat(cache_path_, "cache-d")));
+}
+
+TEST_F(FileSystemHttpCacheTestWithNoDefaultCache, EvictsOldestFilesUntilUnderSizeLimit) {
+  const std::string file_contents = "XXXXX";
+  const std::string large_file_contents = "XXXXXXXXXXXX";
+  const uint64_t max_size = large_file_contents.size();
+  ConfigProto cfg = testConfig();
+  cfg.mutable_max_cache_size_bytes()->set_value(max_size);
+  env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-a"), file_contents, true);
+  env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-b"), file_contents, true);
+  {
+    // TODO(ravenblack): replace this with backdating the files when that's possible.
+    absl::Mutex mu;
+    absl::MutexLock lock(&mu);
+    auto cond = []() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu) { return false; };
+    mu.AwaitWithTimeout(absl::Condition(&cond), absl::Seconds(1));
+  }
+  cache_ = std::dynamic_pointer_cast<FileSystemHttpCache>(
+      http_cache_factory_->getCache(cacheConfig(cfg), context_));
+  waitForEvictionThreadIdle();
+  env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-c"), large_file_contents, true);
+  EXPECT_EQ(cache_->stats().size_bytes_.value(), file_contents.size() * 2);
+  EXPECT_EQ(cache_->stats().size_count_.value(), 2);
+  cache_->trackFileAdded(large_file_contents.size());
+  waitForEvictionThreadIdle();
+  EXPECT_EQ(cache_->stats().size_bytes_.value(), large_file_contents.size());
+  EXPECT_EQ(cache_->stats().size_count_.value(), 1);
```

Comment:
I see you have a test for it but I am failing to see where size_count_ gets decremented. Can you help me out?

One theory I have is that there's a subsequent GC call which resets size_count_ on cache_eviction_thread.cc:123.

This reminds me, can you also keep counters for how many times eviction runs and test that here?

Reply:
The sequence of events here is:
1. trackFileAdded increments the count to 3, and signals eviction thread.
2. eviction thread counts the files, and sets the count to 3 again.
3. eviction thread evicts files, calls trackFileRemoved, decrementing the count (twice).

Added eviction_runs to stats, and included it in the tests.

Reply:
got it -- I had missed that in my scan of the changes somehow.